index.rst

```diff
@@ -6,24 +6,499 @@
+
+.. Note::
+
+    A method for clearing memory of, for example, large arrays that are not going to be used further on in the notebook is in development.
+
+
+Assert statements
+-----------------
+
+Where essential, or where a very specific value is expected, the ``assert`` command can be used to demonstrate to users that a condition is true.
+
+For example, ``assert`` statements can be used to confirm that service objects like TAP are not ``None`` or ``null`` before moving on and using that instance,
+or to check that values meet expectations (e.g., total rows returned from a query).
+
+However, take care not to use ``assert`` statements when, e.g., querying dynamic (prompt) datasets.
+Consider more pedagogical alternatives when possible (e.g., printing schema columns would also fail if the TAP service was not instantiated).
+
+
+Known warnings
+--------------
+
+If a code cell produces a warning which is known and it should be ignored, the preferred method is to add a markdown cell
+*before* the code cell which produces the warning, to tell the user it is acceptable to ignore.
+
+.. Note::
+
+   This is not preferred because ignore categories of warnings can allow real issues to go unnoticed.
+   The appropriate use-cases for ignoring categories of warnings (below) is still under consideration.
+
+::
+
+  warnings.simplefilter("ignore", category=UserWarning)
+
+
+Markdown style
+--------------
+
+Any references to variables used in code cells or any code commands should be in ``monospaced font``.
+
+Use of indented text should be limited to warnings and notices, e.g., ``> **Warning:** the following cell...``.
+
+.. Note::
+
+   The most appropriate narrative voice for tutorials is still under consideration.
+   Adopt your own style and apply it consistently throughout the notebook.
+
+
+Code cell comments
+------------------
+
+Markdown cells are the preferred way to provide descriptive text.
+Avoid using comments within a code cell as documentation.
+
+
+Code cell style standard PEP8
+-----------------------------
+
+``PEP8`` is the standard, and ``flake8`` is a tool to ensure compliance with the standards.
+
+Use ``flake8`` to ensure notebook code conforms to  `PEP 8 -- Style Guide for Python Code <https://www.python.org/dev/peps/pep-0008/>`_, with a few exceptions.
+
+Notebook tutorial developers must install the following packages locally in their home directory:
+
+::
+
+  pip install --user flake8-nb
+  pip install --user pycodestyle_magic
+
+It is known that the most up-to-date version of ``flake8`` has some issues.
+If errors are encountered such as ``AttributeError: '_io.StringIO' object has no attribute 'buffer'``,
+force-downgrade ``flake8`` from version ``4.0.1`` to ``3.9.2`` with ``pip install flake8=3.9.2``.
```

Comment:
I think ``pip install flake8=3.9.2`` should instead be ``pip install --user flake8==3.9.2``? (requires the "--user" and the "==")

Reply:
fixed